### PR TITLE
[R2] CommitLag Metric + Follower State

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose
Add two new metrics to the log status:
- CommitLag: difference in miliseconds of the insert timepoint of the first uncommitted entry in the log and now.
- Follower State: Reports if the follower is `up-to-date`, `error-backoff` (including retryCount and time until retry), `request-in-flight` (including time passed since request was sent)

### CODE REVIEW

File: arangod/Replication2/ReplicatedLog/types.cpp [hunk 1]
```diff
@@ -111,3 +110,71 @@ auto arangodb::replication2::replicated_log::to_string(AppendEntriesErrorReason
...
+}
+
+auto FollowerState::withRequestInFlight(double duration) noexcept -> FollowerState {
+  return FollowerState(std::in_place, RequestInFlight{duration});
+}
+
+constexpr const static std::string_view upToDateString = "up-to-date";
+constexpr const static std::string_view errorBackoffString = "error-backoff";
+constexpr const static std::string_view requestInFlightString =
+    "request-in-flight";
```
Comment (goedderz) [inline arangod/Replication2/ReplicatedLog/types.cpp:130]:
```suggestion
constexpr static std::string_view upToDateString = "up-to-date";
constexpr static std::string_view errorBackoffString = "error-backoff";
constexpr static std::string_view requestInFlightString = "request-in-flight";
```

**Review by goedderz — APPROVED** [review]

File: arangod/Replication2/ReplicatedLog/LogStatus.h [hunk 2]
```diff
@@ -33,17 +33,20 @@
...
-struct LeaderStatus {
-  struct FollowerStatistics : LogStatistics {
-    AppendEntriesErrorReason lastErrorReason;
-    double lastRequestLatencyMS;
-    void toVelocyPack(velocypack::Builder& builder) const;
-    static auto fromVelocyPack(velocypack::Slice slice) -> FollowerStatistics;
-  };
+struct FollowerStatistics : LogStatistics {
+  AppendEntriesErrorReason lastErrorReason;
+  double lastRequestLatencyMS;
```
Comment (mpoeter) [inline arangod/Replication2/ReplicatedLog/LogStatus.h:38]: How about we make this a `std::chrono::duration<double, std::milli>`?

File: arangod/Replication2/ReplicatedLog/types.h [hunk 3]
```diff
@@ -40,6 +41,33 @@ class Slice;
 
 namespace arangodb::replication2::replicated_log {
 
+struct FollowerState {
+
+  struct UpToDate {};
+  struct ErrorBackoff {
+    double durationMS;
```
Comment (mpoeter) [inline arangod/Replication2/ReplicatedLog/types.h:48]: How about we make this a `std::chrono::duration<double, std::milli>`?

File: arangod/Replication2/ReplicatedLog/types.h [hunk 4]
```diff
@@ -40,6 +41,33 @@ class Slice;
...
+    std::size_t retryCount;
+  };
+  struct RequestInFlight {
+    double durationMS;
```
Comment (mpoeter) [inline arangod/Replication2/ReplicatedLog/types.h:52]: How about we make this a `std::chrono::duration<double, std::milli>`?

File: arangod/Replication2/ReplicatedLog/LogStatus.h [hunk 5]
```diff
@@ -33,17 +33,20 @@
...
+  void toVelocyPack(velocypack::Builder& builder) const;
+  static auto fromVelocyPack(velocypack::Slice slice) -> FollowerStatistics;
+};
 
+struct LeaderStatus {
   LogStatistics local;
   LogTerm term;
   std::unordered_map<ParticipantId, FollowerStatistics> follower;
+  // now() - insertTP of last uncommitted entry
+  double commitLagMS;
```
Comment (mpoeter) [inline arangod/Replication2/ReplicatedLog/LogStatus.h:49]: How about we make this a `std::chrono::duration<double, std::milli>`?

**Review by mpoeter — APPROVED** [review]

**Pull request merged**